src/Microsoft.IdentityModel.Tokens/TokenValidationResult.cs

```diff
@@ -14,20 +16,33 @@ namespace Microsoft.IdentityModel.Tokens
     /// </summary>
     public class TokenValidationResult
     {
-        private Lazy<IDictionary<string, object>> _claims;
-        private ClaimsIdentity _claimsIdentity;
+        private readonly TokenValidationParameters _validationParameters;
+        private readonly TokenHandler _tokenHandler;
+
         private Exception _exception;
         private bool _hasIsValidOrExceptionBeenRead = false;
         private bool _isValid = false;
-        private TokenValidationParameters _validationParameters;
-        private TokenHandler _tokenHandler;
+
+        // Fields lazily initialized in a thread-safe manner. _claimsIdentity is protected by the _claimsIdentitySyncObj
+        // lock, and since null is a valid initialized value, _claimsIdentityInitialized tracks whether or not it's valid.
+        // _claims is constructed by reading the data from the ClaimsIdentity and is synchronized using Interlockeds
+        // to ensure only one dictionary is published in the face of concurrent access (but if there's a race condition,
+        // multiple dictionaries could be constructed, with only one published for all to see). Simiarly, _propertyBag
+        // is initalized with Interlocked to ensure only a single instance is published in the face of concurrent use.
+        // _claimsIdentityInitialized only ever transitions from false to true, and is volatile to reads/writes are not
+        // reordered relative to the other operations. The rest of the objects are not because the .NET memory model
+        // guarantees object writes are store releases and that reads won't be introduced.
+        private volatile bool _claimsIdentityInitialized;
```

Comment:
apologies if I'm missing something here, but why does this need to be volatile?

Reply:
If it's not volatile, then the write to set it to true can effectively move to before the write that initializes the thing being represented, e.g.
```C#
_value = value;
_valueInitialized = true;
```
could be executed as if it were:
```C#
_valueInitialized = true;
_value = value;
```
Then another thread reading _valueInitialized could see it as true before _value was fully set.  This won't happen on x86/x64, which have a strong memory model, but it could happen on ARM which has a weaker memory model.